### PR TITLE
Move "current-user" settings out of `config.json`

### LINKED ISSUE #1721: User is forced to `azd login` again after `azd config reset`
As the title says. if the user runs `azd config reset`, the user is logged out.

1. `azd login` once
1. `azd config reset`
2. `azd init` 

Expected:
`azd init` should succeed.

Actual:
`azd init` reports `Error: not logged in, run `azd login` to login`

### PR DESCRIPTION
**Pull request by ellismg**
Previously, the identity of currently logged in user (either the home account id, when logged in as a user, or the client and tenant id when logged in as a service principal) as stored in user's `config.json`.

A downside of storing this information in `config.json` is that when a user did `azd config reset`, it would "log out" the current user, which is very surprising behavior. In addition, anyone looking at `config.json` would see properties that were much more like state of the auth subsystem instead of actual config values that a user would want to change.

So, this change moves this information out of `config.json` and instead stores it in `auth.json` (a new file) which is also stored in `~/.azd`.

On first use, we'll migrate the relevent properties from `config.json` into `auth.json`.

Fixes #1721

### CODE REVIEW

File: cli/azd/pkg/auth/manager.go [hunk 1]
```diff
@@ -590,7 +611,68 @@ func (m *Manager) saveUserProperties(user *userProperties) error {
...
+
+	return authCfg, nil
+}
+
+func (m *Manager) saveAuthConfig(c config.Config) error {
+	cfgPath, err := config.GetUserConfigDir()
+	if err != nil {
+		return fmt.Errorf("getting user config dir: %w", err)
+	}
+
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager.go:672]: Do you think it's worth filing an issue to make this an initialization `panic`? And we cache the value throughout app lifetime.

File: cli/azd/pkg/auth/manager.go [hunk 2]
```diff
@@ -590,7 +611,68 @@ func (m *Manager) saveUserProperties(user *userProperties) error {
...
+	curUserData, has := cfg.Get(cCurrentUserKey)
+	if !has {
+		return config.NewConfig(map[string]any{}), nil
+	}
+
+	if err := cfg.Unset(cCurrentUserKey); err != nil {
+		return nil, err
+	}
+
+	authCfg := config.NewConfig(nil)
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager.go:651]: What's the difference between `config.NewConfig(nil)` and `config.NewConfig(map[string]any{})` set on line 644?

stylistic nit: I could see how we could construct the empty `authCfg` on line 641, and set the `currentUserData` value if needed.

File: cli/azd/pkg/auth/manager.go [hunk 3]
```diff
@@ -590,7 +611,68 @@ func (m *Manager) saveUserProperties(user *userProperties) error {
...
+	if err != nil {
+		return nil, fmt.Errorf("reading user config: %w", err)
+	}
+
+	curUserData, has := cfg.Get(cCurrentUserKey)
+	if !has {
+		return config.NewConfig(map[string]any{}), nil
+	}
+
+	if err := cfg.Unset(cCurrentUserKey); err != nil {
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager.go:647]: Can we unset this after we've migrated the old value?

File: cli/azd/pkg/auth/manager_test.go [hunk 4]
```diff
@@ -246,25 +254,76 @@ func TestLoginDeviceCode(t *testing.T) {
...
+		publicClient:      &mockPublicClient{},
+	}
+
+	cfg, err := m.readAuthConfig()
+	require.NoError(t, err)
+
+	properties, err := readUserProperties(cfg)
+	require.NoError(t, err)
+	require.NotNil(t, properties.HomeAccountID)
+	require.Equal(t, "homeAccountID", *properties.HomeAccountID)
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager_test.go:282]: nit here: Can we also validate `userCfg` no longer contains the properties? (Or maybe that's not necessary)

File: cli/azd/pkg/auth/manager.go [hunk 5]
```diff
@@ -590,7 +611,68 @@ func (m *Manager) saveUserProperties(user *userProperties) error {
...
+		return nil, err
+	}
+
+	if err := m.configManager.Save(authCfg, authCfgFile); err != nil {
+		return nil, err
+	}
+
+	if err := m.saveAuthConfig(authCfg); err != nil {
+		return nil, err
+	}
```
Comment (weikanglim) [inline cli/azd/pkg/auth/manager.go:662]: Could see us maybe going either way, but I think we value data persistence over failing cleanup. Maybe reverse the older of save. Save new config first, and then save old one with the value unset?

**Review by weikanglim — APPROVED** [review]
LGTM. Minor comments about the persistence logic, but they can be addressed rather easily if applicable.

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/pkg/auth/manager.go (hunk 3 above)
Comment (ellismg) [reply cli/azd/pkg/auth/manager.go:647]: > Nvm, I think unset changes the in-memory object but doesn't persist. Feel free to close this if that's your understanding as well.

Yes, that's my understanding as well.

>  It's weird that unset returns error

Yeah, it is.  I guess in some ways you could imagine trying to unset an invalid key like `foo..bar` or something and that leading to an error.

File: cli/azd/pkg/auth/manager.go (hunk 1 above)
Comment (ellismg) [reply cli/azd/pkg/auth/manager.go:672]: Opened #2078 to track.

**Pull request merged**